apps/shared/avifpng.c

```diff
@@ -17,6 +18,106 @@ typedef png_bytep png_iccp_datap;
 typedef png_charp png_iccp_datap;
 #endif
 
+// Converts a hexadecimal string which contains 2-byte character representations of hexadecimal values to raw data.
+// hexString may contain values consisting of [A-F][a-f][0-9] in pairs, e.g., 7af2..., separated by any number of newlines.
+// On success the bytes are filled and AVIF_TRUE is returned.
+// AVIF_FALSE is returned if fewer than expectedLength hexadecimal pairs are converted.
+static avifBool avifHexStringToBytes(const char * hexString, size_t expectedLength, avifRWData * bytes)
+{
+    avifRWDataRealloc(bytes, expectedLength);
+    const char * src = hexString;
+    uint8_t * dst = bytes->data;
+    size_t actualLength = 0;
+    for (; (actualLength < expectedLength) && (*src != '\0'); ++actualLength, ++dst) {
+        while (*src == '\n') {
+            ++src;
+        }
+        const char hexValue[3] = { src[0], src[1], '\0' };
+        src += 2;
+        char * end;
+        bytes->data[actualLength] = (uint8_t)strtol(hexValue, &end, 16);
+        if (end != &hexValue[2]) {
+            avifRWDataFree(bytes);
+            fprintf(stderr, "Exif extraction failed: invalid token at " AVIF_FMT_ZU "\n", actualLength);
+            return AVIF_FALSE;
+        }
+    }
+
+    if (actualLength != expectedLength) {
+        avifRWDataFree(bytes);
+        fprintf(stderr, "Exif extraction failed: expected " AVIF_FMT_ZU " tokens but got " AVIF_FMT_ZU "\n", expectedLength, actualLength);
+        return AVIF_FALSE;
+    }
+    return AVIF_TRUE;
+}
+
+// Parses the raw profile string of profileLen characters and extracts the payload.
+static avifBool avifCopyRawProfile(const char * profile, size_t profileLen, avifRWData * payload)
+{
+    if (!profile || (profileLen < 3)) {
+        fprintf(stderr, "Exif extraction failed: empty profile\n");
+        return AVIF_FALSE;
+    }
+
+    // ImageMagick formats 'raw profiles' as "\n<name>\n<length>(%8lu)\n<hex payload>\n".
+    const char * src = profile;
+    if (*src != '\n') {
+        fprintf(stderr, "Exif extraction failed: malformed raw profile, expected '\\n' but got '\\x%.2X'\n", *src);
+        return AVIF_FALSE;
+    }
+    ++src;
+    size_t numReadCharacters = 1;
+    // Skip the profile name and extract the length.
+    for (; (*src != '\0') && (*src != '\n'); ++src, ++numReadCharacters) {
+        if (numReadCharacters == profileLen) {
+            fprintf(stderr, "Exif extraction failed: truncated raw profile of size " AVIF_FMT_ZU "\n", profileLen);
+            return AVIF_FALSE;
+        }
+    }
+    char * end;
+    const long expectedLength = strtol(src, &end, 10);
```

Comment:
It seems that we should check if `errno` is equal to `ERANGE` after this `strtol()` call.